Document daml sandbox and using canton-console (#21293)
digital-asset/daml: sdk/docs/manually-written/sdk/component-howtos/application-development/daml-sandbox.rst

.. _component-howtos-application-development-daml-sandbox:

.. _sandbox-manual:

Daml Sandbox
============

The Daml Sandbox, or Sandbox for short, is a program that enables rapid application prototyping by running a simple Canton ledger with one participant.

Install
-------

Install the Daml Sandbox by :brokenref:`installing the Daml Assistant <Install section of Daml Assistant>`.

Configure
---------

> To configure the Daml Sandbox, we use the command line.

Reviewer: ```suggestion
To configure the Daml Sandbox, use the command line.
```

        --dar PATH               DAR file to upload to sandbox
        --shutdown-stdin-close   Shut down when stdin is closed, disabled by default
        -h,--help                Show this help text

Any unrecognized command-line arguments will be treated as arguments to be
supplied to the underlying Canton JAR. Display these options by running ``daml sandbox --canton-help``
in your terminal.

Canton configuration
^^^^^^^^^^^^^^^^^^^^

Behind the scenes, Daml Sandbox runs an underlying Canton ledger with a default
configuration file to initialize a participant named ``sandbox``, a sequencer
named ``sequencer1``, and a mediator named ``mediator1``.

Configure the underlying Canton ledger further in one of two ways:

* Specify additional configuration file(s) to apply on top of the default
  configuration using ``--config filepath``. If several configuration files
  assign to the same key, the last value is taken.

* Set the value of a specific key with ``-C key=value``. For example, to
  override the ledger-api port of the sandbox participant, the command would be:

  .. code-block:: none

        daml sandbox -C canton.participants.sandbox.ledger-api.port=9999

Operate
-------

> There are two ways to start running the Daml Sandbox:

Reviewer: ```suggestion
There are two ways to run the Daml Sandbox:
```

against it:

.. code-block:: none

    $ daml ledger upload-dar --host localhost --port 6865 <path to DAR>
    $ daml script --ledger-host localhost --port 6865 --dar <path to DAR> --script-name <script name in DAR>

Because ``daml sandbox`` is a Canton instance, all documentation for using Canton applies.

Connecting to Sandbox's console
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Once you have a Sandbox running locally (i.e. after running ``daml start`` or ``daml sandbox``)
you may connect to Sandbox remotely by running the ``daml canton-console``
command in a separate terminal:

.. code-block:: none

    $ daml canton-console
       _____            _
      / ____|          | |
     | |     __ _ _ __ | |_ ___  _ __
     | |    / _` | '_ \| __/ _ \| '_ \
     | |___| (_| | | | | || (_) | | | |
      \_____\__,_|_| |_|\__\___/|_| |_|

      Welcome to Canton!
      Type `help` to get started. `exit` to leave.

    @

You can quit the session by running the ``exit`` command.

Built-in documentation
""""""""""""""""""""""

The Canton console comes with built-in documentation. You
can use the ``help`` command to get online documentation for top-level commands. Many objects in the

> console also have further built-in help that you can access by invoking the ``help`` method on them.

Reviewer: ```suggestion
console also have built-in help that you can access by invoking the ``help`` method on them.
```

For example, you can ask for help on the ``health`` object by typing:

.. code-block:: scala

  health.help

> Or go more in depth about specific items within that object as in the following example:

Reviewer: ```suggestion
Or go more in-depth about specific items within that object, as in the following example:
```

.. code-block:: scala

  health.help("status")

Interact with the Sandbox
"""""""""""""""""""""""""

One of the objects available in the Canton console represents the Sandbox itself. The object is called
``sandbox`` and you can use it to interact with the Sandbox. For example, you can list the DARs loaded
on the Sandbox by running the following command:

.. code-block:: scala

  sandbox.dars.list()

Among the various features available as part of the console, you can manage parties and packages,

> check the health of the Sandbox, perform pruning operations and more. Consult the built-in documentation mentioned

Reviewer: ```suggestion
check the health of the Sandbox, perform pruning operations, and more. Consult the built-in documentation mentioned
```

         // APIs. If the server certificate is trusted by the trust-collection, then you can
         // just use the server certificates. Otherwise, you need to create separate ones.
         admin-client {
           cert-chain-file = "./tls/admin-client.crt"
           private-key-file = "./tls/admin-client.pem"
         }
       }
     }
   }

Troubleshoot
------------

Failed to bind to address
^^^^^^^^^^^^^^^^^^^^^^^^^

By default, Daml Sandbox reserves five ports for its Canton services:

* ``6865`` for the participant's Ledger API
* ``6866`` for the participant's Admin API
* ``6867`` for the sequencer's public API
* ``6868`` for the sequencer's admin API
* ``6869`` for the mediator's admin API

The Sandbox will also bind to the port specified in the ``--json-api-port``, if
any.

When one of these ports is already used by an existing process, Sandbox will
emit an error that contains the following text:

.. code-block:: none

   Failed to bind to address /127.0.0.1:<port number>

This is most commonly either caused by an existing process that is already
listening on that port, or if you do not have the permissions to bind to that
address.

> On Linux, the ``lsof -n -i`` command can help you discover lists what processes

Reviewer: ```suggestion
On Linux, the ``lsof -n -i`` command can help you discover lists that
```

Reviewer: I'm not sure if this mangles the intent of the sentence, but the way it was written didn't make sense to me.

Author: I miswrote that sentence by mashing two wordings together - I've fixed the grammar on it.

are already listening to a port. For example, if an existing Java program is
already listening to 6865, ``lsof`` would look as follows:

.. code-block:: none

      $ lsof -n -i
      ...
      java       707977 username       77u  IPv6 67556378      0t0  TCP 127.0.0.1:6865 (LISTEN)
      ...

If killing the existing process isn't an option, or if you don't have the
permission to bind to a given port, you can reconfigure the ports of a given

> node using the top-level options described in.

Reviewer: ```suggestion
node using the top-level options described below:
```

* Use ``--port=<port>`` to override binding to ``6865``
* Use ``--admin-api-port=<port>`` to override binding to ``6866``
* Use ``--sequencer-public-port=<port>`` to override binding to ``6867``
* Use ``--sequencer-admin-port=<port>`` to override binding to ``6868``
* Use ``--mediator-admin-port=<port>`` to override binding to ``6869``
* Use ``--json-api-port`` to change the port to which the JSON API binds.

SDK not installed
^^^^^^^^^^^^^^^^^

If the ``daml.yaml`` file of the project you are currently in specifies a
version of the Daml SDK that is not installed, you may get the following error
message:

.. code-block:: none

      SDK not installed. Cannot run command without SDK.

> In order to fix this, either

Reviewer: ```suggestion
To fix this, you can:
```
